src/js/d3ViewComponent.js

```diff
@@ -114,11 +139,93 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
         return output;
     };
 
+    // Given a data key used to maintain object constancy in D3, a DOM
+    // element with a unique ID and the component, updates the key's
+    // value (an array of IDs) to include that ID
+    floe.d3ViewComponent.addElementIdToDataKey = function (d3Key, idToAdd, that) {
+        var keyPath = ["dataKeys", d3Key, idToAdd];
+        that.applier.change(keyPath, true);
+    };
+
+    // Corresponding "remove" functionality to addElementIdToDataKey
+    floe.d3ViewComponent.removeElementIdFromDataKey = function (d3Key, idToRemove, that) {
+        var keyPath = ["dataKeys", d3Key, idToRemove];
+        that.applier.change(keyPath, false, "DELETE");
+    };
+
+    // Given an object "elements" with element IDs as keys, returns a
+    // joined string of IDs suitable for use as a jQuery selector to select all
+    // those IDs
+    floe.d3ViewComponent.getElementIdsAsSelector = function (elementIds) {
+        if (fluid.isPlainObject(elementIds)) {
+            var keys = Object.keys(elementIds);
+            var elemIdCollectionWithPreface = fluid.transform(keys, function (elemId) {
+                    return "#" + elemId;
+                });
+            var keyedElements = elemIdCollectionWithPreface.join(", ");
+            return keyedElements;
+        }
+    };
+
+    // Given an array of D3 data keys, returns all affiliated D3-bound elements
+    // using the model's dataKeys information
+    floe.d3ViewComponent.getElementsByDataKeys = function (dataKeys, that) {
+        var matchedElements = [];
+        fluid.each(dataKeys, function (dataKey) {
+            var elementIds = that.model.dataKeys[dataKey];
+            var selector = floe.d3ViewComponent.getElementIdsAsSelector(elementIds);
+            matchedElements.push(selector);
+        });
+        var elements = $(matchedElements.join(","));
+        return elements;
+    };
+
+    // Given a D3 data key, returns all D3-bound elements that aren't associated
+    // with that key
+
+    floe.d3ViewComponent.getElementsNotMatchingDataKey = function (dataKey, that) {
+        var dataKeys = fluid.copy(that.model.dataKeys);
+
+        fluid.remove_if(dataKeys, function (currentObj, currentKey) {
+            return currentKey === dataKey;
+        });
+        return floe.d3ViewComponent.getElementsByDataKeys(Object.keys(dataKeys), that);
+    };
+
+    // Given a selection of D3 elements, an ID and a CSS class, turns that
+    // class on for any elements matching the ID and makes sure it's turned off
+    // for any elements not matching it
+    floe.d3ViewComponent.toggleCSSClassByDataId = function (id, toggleClass, that) {
+        // Get all D3-bound elements
+        var allElements = floe.d3ViewComponent.getElementsByDataKeys(Object.keys(that.model.dataKeys), that);
+
+        allElements.each(function (idx, elem) {
+            var dataId = elem.__data__.id;
+            if (id === dataId) {
```

Comment:
line 211-219 can be simplified to:
```
var shouldHighlight = id === dataId ? true : false;
$(elem).toggleClass(cssClass, shouldHighlight);
elem.classList.[shouldHighlight ? "add" : "remove"](toggleClass);
```

I renamed ```toggleClass``` to ```cssClass``` to avoid the confusion with the in-use jQuery ```toggleClass()``` function.